### PR TITLE
[MM-32622] Remove app.WaitForChannelMembership()

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR removes `app.WaitForChannelMembership()`. To eliminate the need to poll the db until a replica has the updated data we pass a context from the API layer and make both `api4.getChannelMember()` and `api4.channelMemberCountsByGroup()` query from the writer node (master). This is to make sure that clients doing subsequent requests in response to either websocket events or API calls will be guaranteed to access up-to-date data.

#### Ticket

https://mattermost.atlassian.net/browse/MM-32622

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: app/context.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (streamer45) [inline app/context.go:1]: This isn't strictly needed. Adding it to avoid coupling the `api4` package with `sqlstore`.

File: app/channel.go [hunk 2]
```diff
@@ -2989,8 +2953,8 @@ func (a *App) ClearChannelMembersCache(channelID string) {
 	}
 }
 
-func (a *App) GetMemberCountsByGroup(channelID string, includeTimezones bool) ([]*model.ChannelMemberCountByGroup, *model.AppError) {
-	channelMemberCounts, err := a.Srv().Store.Channel().GetMemberCountsByGroup(channelID, includeTimezones)
+func (a *App) GetMemberCountsByGroup(ctx context.Context, channelID string, includeTimezones bool) ([]*model.ChannelMemberCountByGroup, *model.AppError) {
+	channelMemberCounts, err := a.Srv().Store.Channel().GetMemberCountsByGroup(ctx, channelID, includeTimezones)
```
Comment (agnivade) [inline app/channel.go:2957]: Was this also needed?

**Review by agnivade — APPROVED** [review]

File: app/channel.go (hunk 2 above)
Comment (streamer45) [reply app/channel.go:2957]: Yes, that's another call made on LDAP enabled servers by the webapp in response to the same event.

**Review by isacikgoz — APPROVED** [review]
Nice one 🚀

**Comment by streamer45** [comment]
@lindalumitchell 

This will require some smoke-test around:

- Creating direct/group channels.
- Adding users to channels.

Specifically it would be good to perform such tests in a HA environment with at least 2 DB nodes (1WR + 1RD) and verify there are no issues (potentially related to replica lag). Feel free to assign someone from the test engineers. I'll also be available to offer support if needed.

**Comment by lindalumitchell** [comment]
@streamer45 my apologies for this one falling through the cracks. Getting some help on it as soon as possible.

**Comment by saturninoabril** [comment]
@streamer45 Please help check the failed build. For testing purposes, I'll be spinning up a remote HA server using the built docker image from this PR.

**Comment by streamer45** [comment]
> @streamer45 Please help check the failed build. For testing purposes, I'll be spinning up a remote HA server using the built docker image from this PR.

There seems to be some [unrelated issue](https://community.mattermost.com/core/pl/rrg53w64ypfxfqyej7o5se535r) going on. I'll ping you when it's resolved.

**Comment by streamer45** [comment]
@saturninoabril 

This should be ready now :)

**Review by saturninoabril — APPROVED** [review]
Thanks @streamer45! Tested and passed.

**Pull request merged**